File: src/Queue/Broker/Redis.php
<?php

namespace Utopia\Queue\Broker;

use Utopia\Queue\Codec;
use Utopia\Queue\Codec\Json;
use Utopia\Queue\Connection;
use Utopia\Queue\Consumer;
use Utopia\Queue\Message;
use Utopia\Queue\Publisher\Synchronous;
use Utopia\Queue\Queue;

class Redis implements Synchronous, Consumer
{
    private const int POP_TIMEOUT = 2;
    private const int RECONNECT_BACKOFF_MS = 100;
    private const int RECONNECT_MAX_BACKOFF_MS = 5_000;

    /** Heartbeat lifetime; handlers refresh it every third of this interval. */
    private const int CLAIM_TTL = 90;

    /** Lock expiry schedules the next fleet sweep; never release it early. */
    private const int REAP_INTERVAL = 60;

    /** Maximum claims requeued per sweep; bounds work, not elapsed time. */
    private const int REAP_LIMIT = 1_000;

    private bool $closed = false;
    /** @var array<string, \Utopia\Queue\Internal\Buffer> */
    private array $settlements = [];

    /**
     * Queues attempted by this broker, keyed by namespace/name pairs.
     *
     * @var array<string, Queue>
     */
    private array $served = [];
    private int $reconnectAttempt = 0;
    private int $reconnectBackoffMs = self::RECONNECT_BACKOFF_MS;
    /**
     * @var (callable(Queue, \Throwable, int, int): void)|null
     */
    private $reconnectCallback;
    /**
     * @var (callable(Queue, int): void)|null
     */
    private $reconnectSuccessCallback;

    public function __construct(
        // Blocking receive loop + claim writes (single caller).
        private readonly Connection $receive,
        // Acks and publishing; wrap in Locking when shared by coroutines.
        private readonly Connection $commands,
        // How an envelope is written and read; Json is what every release so
        // far has put on the list. See Codec\Compat before changing it on a
        // queue that already holds messages.
        private readonly Codec $codec = new Json(),
        // Minimum publish age for recovery without a live heartbeat.
        // Keep conservative until all workers and adapters heartbeat.
        private readonly int $reapAfter = 90_000,
    ) {}

    public function setReconnectCallback(?callable $callback): self
    {
        $this->reconnectCallback = $callback;

        return $this;
    }

    public function setReconnectSuccessCallback(?callable $callback): self
    {
        $this->reconnectSuccessCallback = $callback;

        return $this;
    }

    public function receive(Queue $queue, int $timeout, int $n = 1): array
    {
        if ($this->isClosed()) {
            return [];
        }

        // An idle queue's stranded claims still need sweeping, so the queue is
        // remembered on the attempt, not on the first message.
        $this->served[serialize([$queue->namespace, $queue->name])] ??= $queue;

        $key = "{$queue->namespace}.queue.{$queue->name}";

        try {
            $token = bin2hex(random_bytes(16));
            $reservation = "{$queue->namespace}.reservations.{$queue->name}.{$token}";
            $keys = [$key, "{$queue->namespace}.reservations.{$queue->name}", $reservation,
                "{$queue->namespace}.processing.{$queue->name}",
                "{$queue->namespace}.stats.{$queue->name}.total",
                "{$queue->namespace}.stats.{$queue->name}.processing"];
            $count = min(self::REAP_LIMIT, max(1, $n));
            $batch = $this->script($this->receive, 'reserve', $keys, [$count, self::CLAIM_TTL, max(0, $timeout), 0]);
            if ($batch === [] && $timeout > 0) {
                // The script registered this reservation before blocking. A crash
                // after the move is recoverable even before PHP sees the reply.
                $raw = $this->receive->rightPopLeftPush($key, $reservation, $timeout);
                if (\is_string($raw)) {
                    $batch = [$raw];
                    if ($count > 1) {
                        $batch = [...$batch, ...$this->script($this->receive, 'reserve', $keys, [$count - 1, self::CLAIM_TTL, 0, 1])];
                    }
                }
            }

            if ($this->reconnectAttempt > 0) {
                $this->triggerReconnectSuccessCallback($queue, $this->reconnectAttempt);
            }

            $this->reconnectBackoffMs = self::RECONNECT_BACKOFF_MS;
            $this->reconnectAttempt = 0;
        } catch (\RedisException|\RedisClusterException $e) {
            if ($this->isClosed()) {
                return [];
            }

            $this->reconnectAttempt++;

            try {
                $this->receive->close();
            } catch (\Throwable) {
            }

            $sleepMs = mt_rand(0, $this->reconnectBackoffMs);
            $this->triggerReconnectCallback($queue, $e, $this->reconnectAttempt, $sleepMs);

            usleep($sleepMs * 1000);
            $this->reconnectBackoffMs = min(self::RECONNECT_MAX_BACKOFF_MS, $this->reconnectBackoffMs * 2);

            return [];
        }

        if ($batch === []) {
            return [];
        }

        return $this->claim($queue, $batch, $token, $reservation);
    }

    /** Finalize recoverable raw reservations without teaching Lua application codecs. */
    private function claim(Queue $queue, array $batch, string $token, string $reservation): array
    {
        $keys = ["{$queue->namespace}.reservations.{$queue->name}", $reservation,
            "{$queue->namespace}.processing.{$queue->name}",
            "{$queue->namespace}.stats.{$queue->name}.total",
            "{$queue->namespace}.stats.{$queue->name}.processing",
            "{$queue->namespace}.poison.{$queue->name}"];
        $messages = $args = $poison = [];
        foreach ($batch as $raw) {
            try {
                $envelope = $this->codec->decode($raw);
                if (!\is_array($envelope) || !isset($envelope['pid'], $envelope['queue'], $envelope['timestamp'])) {
                    throw new \UnexpectedValueException('Invalid queue envelope');
                }
                $envelope['timestamp'] = (int) $envelope['timestamp'];
                $message = new Message($envelope);
            } catch (\Throwable) {
                $poison[] = $raw;
                continue;
            }
            $message->setReceipt($token);
            $messages[] = $message;
            $pid = $message->getPid();
            $keys[] = "{$queue->namespace}.jobs.{$queue->name}.{$pid}";
            $keys[] = "{$queue->namespace}.claims.{$queue->name}.{$pid}";
            $keys[] = "{$queue->namespace}.owners.{$queue->name}.{$pid}";
            $args[] = $raw;
            $args[] = $pid;
        }
        $this->script($this->receive, 'claim', $keys, [self::CLAIM_TTL, $token, \count($messages), ...$args, ...$poison]);
        return $messages;
    }

    public function commit(Queue $queue, Message $message): void
    {
        $this->settle($queue, $message, 'commit');
    }

    public function reject(Queue $queue, Message $message): void
    {
        $this->settle($queue, $message, 'reject');
    }

    /** @param 'commit'|'reject'|'release' $operation */
    private function settle(Queue $queue, Message $message, string $operation): void
    {
        $pid = $message->getPid();
        $outcome = $operation === 'commit' ? 'success' : 'failed';
        $list = $operation === 'release' ? 'queue' : ($message->isTerminal() ? 'dead' : 'failed');
        $this->settlements[$queue->namespace] ??= new \Utopia\Queue\Internal\Buffer(function (array $requests, callable $resolved): void {
            $keys = $args = [];
            foreach ($requests as [$requestKeys, $requestArgs]) {
                array_push($keys, ...$requestKeys);
                array_push($args, ...$requestArgs);
            }
            foreach ($this->script($this->commands, 'settle', $keys, $args) as $index => $result) {
                $resolved($index, $result === false ? new \RedisException('Queue settlement failed') : $result);
            }
        });
        $result = $this->settlements[$queue->namespace]->request([[
            "{$queue->namespace}.claims.{$queue->name}.{$pid}",
            "{$queue->namespace}.jobs.{$queue->name}.{$pid}",
            "{$queue->namespace}.processing.{$queue->name}",
            "{$queue->namespace}.stats.{$queue->name}.processing",
            "{$queue->namespace}.stats.{$queue->name}.{$outcome}",
            "{$queue->namespace}.{$list}.{$queue->name}",
            "{$queue->namespace}.owners.{$queue->name}.{$pid}",
        ], [$message->getReceipt() ?? '', $pid, $operation, $queue->jobTtl]]);
        if ($result !== 1) {
            throw new \RuntimeException('Queue delivery is no longer owned by this consumer');
        }
    }

    /** Return prefetched work that never entered its handler, without counting a failure. */
    public function release(Queue $queue, Message ...$messages): void
    {
        foreach (array_reverse($messages) as $message) {
            $this->settle($queue, $message, 'release');
        }
    }

    public function extend(Queue $queue, Message ...$messages): void
    {
        if ($messages === []) {
            return;
        }
        $keys = $tokens = [];
        foreach ($messages as $message) {
            $keys[] = "{$queue->namespace}.claims.{$queue->name}.{$message->getPid()}";
            $keys[] = "{$queue->namespace}.owners.{$queue->name}.{$message->getPid()}";
            $tokens[] = $message->getReceipt() ?? '';
        }
        foreach (array_chunk($keys, self::REAP_LIMIT * 2) as $index => $chunk) {
            $this->script($this->commands, 'extend', $chunk, [self::CLAIM_TTL, ...\array_slice($tokens, $index * self::REAP_LIMIT, self::REAP_LIMIT)]);
        }
    }

    private function script(Connection $connection, string $name, array $keys, array $args): mixed
    {
        static $scripts = [];
        $script = $scripts[$name] ??= file_get_contents(__DIR__ . '/Redis/' . $name . '.lua');
        return $connection->execute($script, $keys, $args);
    }

    /** Refresh often enough to tolerate two missed beats. */
    public function extendInterval(): float
    {
        return self::CLAIM_TTL / 3;
    }

    /** Recover stranded claims from served queues, once per fleet interval. */
    public function maintain(): void
    {
        foreach ($this->served as $queue) {
            if (!$this->commands->setNotExists("{$queue->namespace}.reap-lock.{$queue->name}", (string) time(), self::REAP_INTERVAL)) {
                continue;
            }

            $registry = "{$queue->namespace}.reservations.{$queue->name}";
            $expired = $this->script($this->commands, 'expired', [$registry], [self::REAP_LIMIT]);
            if ($expired !== []) {
                $this->script($this->commands, 'recover', [$registry, "{$queue->namespace}.queue.{$queue->name}", ...$expired], [self::REAP_LIMIT]);
            }
            $this->reap($queue, $this->reapAfter, limit: self::REAP_LIMIT, scan: self::REAP_LIMIT * 2);
        }
    }

    /** Idle resource hook used by Utopia\Pools\Pool::maintain(). */
    public function tick(): void
    {
        $this->maintain();
    }

    public function close(): void
    {
        $this->closed = true;
    }

    /** @phpstan-impure close() flips this from another coroutine mid-receive(). */
    private function isClosed(): bool
    {
        return $this->closed;
    }

    private function triggerReconnectCallback(Queue $queue, \Throwable $error, int $attempt, int $sleepMs): void
    {
        if (!\is_callable($this->reconnectCallback)) {
            return;
        }

        try {
            ($this->reconnectCallback)($queue, $error, $attempt, $sleepMs);
        } catch (\Throwable) {
        }
    }

    private function triggerReconnectSuccessCallback(Queue $queue, int $attempts): void
    {
        if (!\is_callable($this->reconnectSuccessCallback)) {
            return;
        }

        try {
            ($this->reconnectSuccessCallback)($queue, $attempts);
        } catch (\Throwable) {
        }
    }

    public function publish(Queue $queue, array $payload): bool
    {
        $key = "{$queue->namespace}.queue.{$queue->name}";
        $envelope = $this->codec->encode($this->envelope($queue, $payload));

        return $this->commands->leftPush($key, $envelope);
    }

    public function publishMany(Queue $queue, array $payloads): bool
    {
        if ($payloads === []) {
            return true;
        }

        $encoded = [];
        foreach ($payloads as $payload) {
            $encoded[] = $this->codec->encode($this->envelope($queue, $payload));
        }

        $key = "{$queue->namespace}.queue.{$queue->name}";

        return $this->commands->leftPushMany($key, $encoded);
    }

    /**
     * @param array<string, mixed> $payload
     * @return array<string, mixed>
     */
    private function envelope(Queue $queue, array $payload): array
    {
        return [
            'pid' => uniqid(more_entropy: true),
            'queue' => $queue->name,
            'timestamp' => time(),
            'payload' => $payload,
        ];
    }

    /**
     * Take all jobs from the failed queue and requeue them.
     *
     * @param int|null $limit The amount of jobs to retry
     * @param int|null $maxAttempts Jobs requeued this many times are parked on
     *        the dead queue instead of looping forever; null retries unbounded.
     * @param int|null $newerThan Only jobs published within this many seconds
     *        are requeued; older ones are parked on the dead queue. Payloads
     *        never expire by default, so without this bound a sweep would
     *        resurrect arbitrarily old work.
     */
    public function retry(Queue $queue, ?int $limit = null, ?int $maxAttempts = null, ?int $newerThan = null): void
    {
        $start = time();
        $processed = 0;

        while ($limit === null || $processed < $limit) {
            $pid = $this->commands->rightPop("{$queue->namespace}.failed.{$queue->name}", self::POP_TIMEOUT);

            // No more jobs to retry
            if ($pid === false) {
                break;
            }

            // The payload expired; nothing left to requeue.
            $job = $this->getJob($queue, $pid);
            if ($job === false) {
                continue;
            }

            // Wrapped around to a job this sweep already requeued: put the
            // claim back and stop.
            if ($job->getTimestamp() >= $start) {
                $this->commands->rightPush("{$queue->namespace}.failed.{$queue->name}", $pid);
                break;
            }

            if (($maxAttempts !== null && $job->getAttempts() >= $maxAttempts)
                || ($newerThan !== null && $job->getTimestamp() < $start - $newerThan)) {
                $this->commands->leftPush("{$queue->namespace}.dead.{$queue->name}", $pid);
                continue;
            }

            $this->requeue($queue, $job);
            $processed++;
        }
    }

    /**
     * Recover claims left between receive() and commit/reject(). Live heartbeats
     * protect running handlers; the publish-age gate protects workers without
     * heartbeats. Keep $olderThan above their longest possible runtime.
     *
     * @param int $olderThan Seconds since publication before a heartbeat-less claim
     *        counts as stale
     * @param int|null $limit Maximum number of claims to requeue
     * @param int|null $maxAttempts Claims requeued this many times are parked
     *        on the dead queue; null reaps unbounded.
     * @param int|null $newerThan Only claims published within this many seconds
     *        are requeued; older ones are parked on the dead queue.
     * @param int|null $scan Maximum claims examined; bounded sweeps share progress.
     * @return int The number of claims requeued
     */
    public function reap(Queue $queue, int $olderThan = 90000, ?int $limit = null, ?int $maxAttempts = null, ?int $newerThan = null, ?int $scan = null): int
    {
        $processing = "{$queue->namespace}.processing.{$queue->name}";
        $now = time();
        $cutoff = $now - $olderThan;
        $requeued = 0;

        $size = $this->commands->listSize($processing);
        if ($size === 0) {
            return 0;
        }

        // Share progress across sweep winners. Retained entries count from the
        // tail; remaining entries bound the cycle so arrivals cannot delay wrapping.
        $key = "{$queue->namespace}.reap-cursor.{$queue->name}";
        $cursor = $scan === null ? [] : explode(':', (string) $this->commands->get($key));
        $retained = (int) ($cursor[0] ?? 0);
        $remaining = (int) ($cursor[1] ?? 0);
        if ($retained < 0 || $retained >= $size || $remaining <= 0) {
            $retained = 0;
            $remaining = $size;
        }
        $remaining = min($remaining, $size - $retained);
        $length = $scan === null ? $size : min(max(1, $scan), $remaining);
        $claims = $this->commands->listRange($processing, $length, max(0, $size - $retained - $length));

        foreach (array_reverse($claims) as $pid) {
            if ($limit !== null && $requeued >= $limit) {
                break;
            }

            $remaining--;
            if (!\is_string($pid)) {
                $retained++;
                continue;
            }

            $ownerKey = "{$queue->namespace}.owners.{$queue->name}.{$pid}";
            $owner = $this->commands->get($ownerKey);
            // Only legacy payloads expire while processing.
            $job = $this->getJob($queue, $pid);
            if ($job === false) {
                if (\is_string($owner)) {
                    throw new \RuntimeException('Queue delivery payload is missing');
                }
                $this->commands->listRemove($processing, $pid);
                continue;
            }

            if ($job->getTimestamp() > $cutoff
                || \is_string($this->commands->get("{$queue->namespace}.claims.{$queue->name}.{$pid}"))) {
                $retained++;
                continue;
            }

            $dead = ($maxAttempts !== null && $job->getAttempts() >= $maxAttempts)
                || ($newerThan !== null && $job->getTimestamp() < $now - $newerThan);
            $moved = $this->script($this->commands, 'reclaim', [
                $ownerKey, "{$queue->namespace}.claims.{$queue->name}.{$pid}",
                "{$queue->namespace}.jobs.{$queue->name}.{$pid}", $processing,
                "{$queue->namespace}.stats.{$queue->name}.processing",
                "{$queue->namespace}." . ($dead ? 'dead' : 'queue') . ".{$queue->name}",
            ], [\is_string($owner) ? $owner : '', $pid, $dead ? '' : $this->retryPayload($queue, $job), $queue->jobTtl]);
            if ($moved && !$dead) {
                $requeued++;
            } elseif (!$moved) {
                $retained++;
            }
        }

        if ($scan !== null) {
            $this->commands->set($key, "{$retained}:{$remaining}");
        }

        return $requeued;
    }

    /**
     * Requeue with a fresh pid and timestamp, carrying the attempt count
     * forward so retry() and reap() can park messages that never succeed.
     */
    private function requeue(Queue $queue, Message $job): void
    {
        $this->commands->leftPush("{$queue->namespace}.queue.{$queue->name}", $this->retryPayload($queue, $job));
    }

    private function retryPayload(Queue $queue, Message $job): string
    {
        $payload = [
            'pid' => uniqid(more_entropy: true),
            'queue' => $queue->name,
            'timestamp' => time(),
            'payload' => $job->getPayload(),
            'attempts' => $job->getAttempts() + 1,
        ];
        return $this->codec->encode($payload);
    }

    private function getJob(Queue $queue, string $pid): Message|false
    {
        $value = $this->commands->get("{$queue->namespace}.jobs.{$queue->name}.{$pid}");

        // Missing or expired jobs come back null or false; anything else is a
        // stored envelope, which only the codec can read.
        if (\is_string($value)) {
            try {
                $value = $this->codec->decode($value);
            } catch (\Throwable) {
                return false;
            }
        }

        return \is_array($value) ? new Message($value) : false;
    }

    public function getQueueSize(Queue $queue, bool $failedJobs = false): int
    {
        $queueName = "{$queue->namespace}.queue.{$queue->name}";
        if ($failedJobs) {
            $queueName = "{$queue->namespace}.failed.{$queue->name}";
        }
        return $this->commands->listSize($queueName);
    }
}

File: 3.x.x/crates-utopia/queue/src/broker/redis.rs
use std::sync::atomic::{AtomicBool, AtomicU32, AtomicU64, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use parking_lot::Mutex;
use serde_json::{json, Value};

use crate::connection::{Connection, StoredValue};
use crate::consumer::Consumer;
use crate::error::QueueError;
use crate::message::Message;
use crate::publisher::Publisher;
use crate::queue::Queue;

const POP_TIMEOUT: i64 = 2;
const RECONNECT_BACKOFF_MS: u64 = 100;
const RECONNECT_MAX_BACKOFF_MS: u64 = 5_000;

/// Redis list broker (processing / failed / dead keys).
///
/// PHP `Utopia\Queue\Broker\Redis`.
#[derive(Clone)]
pub struct Redis {
    receive: Arc<dyn Connection>,
    commands: Arc<dyn Connection>,
    state: Arc<RedisState>,
}

struct RedisState {
    closed: AtomicBool,
    reconnect_attempt: AtomicU32,
    reconnect_backoff_ms: AtomicU64,
    reconnect_callback: Mutex<Option<ReconnectCallback>>,
    reconnect_success_callback: Mutex<Option<ReconnectSuccessCallback>>,
}

/// `(queue, error, attempt, sleep_ms)`
pub type ReconnectCallback = Arc<dyn Fn(&Queue, &QueueError, u32, u64) + Send + Sync>;
/// `(queue, attempts)`
pub type ReconnectSuccessCallback = Arc<dyn Fn(&Queue, u32) + Send + Sync>;

impl std::fmt::Debug for Redis {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Redis")
            .field("closed", &self.state.closed.load(Ordering::Relaxed))
            .finish_non_exhaustive()
    }
}

impl Redis {
    pub fn new(receive: impl Connection + 'static, commands: impl Connection + 'static) -> Self {
        Self::from_arcs(Arc::new(receive), Arc::new(commands))
    }

    pub fn from_arcs(receive: Arc<dyn Connection>, commands: Arc<dyn Connection>) -> Self {
        Self {
            receive,
            commands,
            state: Arc::new(RedisState {
                closed: AtomicBool::new(false),
                reconnect_attempt: AtomicU32::new(0),
                reconnect_backoff_ms: AtomicU64::new(RECONNECT_BACKOFF_MS),
                reconnect_callback: Mutex::new(None),
                reconnect_success_callback: Mutex::new(None),
            }),
        }
    }

    pub fn set_reconnect_callback(&self, callback: Option<ReconnectCallback>) -> &Self {
        *self.state.reconnect_callback.lock() = callback;
        self
    }

    pub fn set_reconnect_success_callback(
        &self,
        callback: Option<ReconnectSuccessCallback>,
    ) -> &Self {
        *self.state.reconnect_success_callback.lock() = callback;
        self
    }

    fn is_closed(&self) -> bool {
        self.state.closed.load(Ordering::SeqCst)
    }

    fn trigger_reconnect(&self, queue: &Queue, error: &QueueError, attempt: u32, sleep_ms: u64) {
        if let Some(cb) = self.state.reconnect_callback.lock().clone() {
            cb(queue, error, attempt, sleep_ms);
        }
    }

    fn trigger_reconnect_success(&self, queue: &Queue, attempts: u32) {
        if let Some(cb) = self.state.reconnect_success_callback.lock().clone() {
            cb(queue, attempts);
        }
    }

    fn get_job(&self, queue: &Queue, pid: &str) -> Result<Option<Message>, QueueError> {
        let key = format!("{}.jobs.{}.{}", queue.namespace, queue.name, pid);
        let value = self.commands.get(&key)?;
        let json = match value {
            Some(StoredValue::Array(v)) => v,
            Some(StoredValue::String(s)) => {
                serde_json::from_str(&s).map_err(|e| QueueError::Other(e.to_string()))?
            }
            None => return Ok(None),
        };
        if json.is_object() {
            Ok(Some(Message::from_value(&json)))
        } else {
            Ok(None)
        }
    }

    fn requeue(&self, queue: &Queue, job: &Message) -> Result<(), QueueError> {
        let payload = json!({
            "pid": uniqid(),
            "queue": queue.name,
            "timestamp": unix_now(),
            "payload": job.get_payload(),
            "attempts": job.get_attempts() + 1,
        });
        self.commands
            .left_push_array(&queue.key("queue"), &payload)?;
        Ok(())
    }
}

impl Publisher for Redis {
    fn enqueue(&self, queue: &Queue, payload: Value, priority: bool) -> Result<bool, QueueError> {
        let envelope = json!({
            "pid": uniqid(),
            "queue": queue.name,
            "timestamp": unix_now(),
            "payload": payload,
        });
        let key = queue.key("queue");
        if priority {
            self.commands.right_push_array(&key, &envelope)
        } else {
            self.commands.left_push_array(&key, &envelope)
        }
    }

    fn retry(
        &self,
        queue: &Queue,
        limit: Option<i64>,
        max_attempts: Option<i64>,
        newer_than: Option<i64>,
    ) -> Result<(), QueueError> {
        let start = unix_now();
        let mut processed = 0i64;
        loop {
            if limit.is_some_and(|l| processed >= l) {
                break;
            }
            let pid = match self.commands.right_pop(&queue.key("failed"), POP_TIMEOUT)? {
                Some(pid) => pid,
                None => break,
            };
            let Some(job) = self.get_job(queue, &pid)? else {
                continue;
            };
            if job.get_timestamp() >= start {
                self.commands.right_push(&queue.key("failed"), &pid)?;
                break;
            }
            if max_attempts.is_some_and(|m| job.get_attempts() >= m)
                || newer_than.is_some_and(|n| job.get_timestamp() < start - n)
            {
                self.commands.left_push(&queue.key("dead"), &pid)?;
                continue;
            }
            self.requeue(queue, &job)?;
            processed += 1;
        }
        Ok(())
    }

    fn get_queue_size(&self, queue: &Queue, failed_jobs: bool) -> Result<i64, QueueError> {
        let key = if failed_jobs {
            queue.key("failed")
        } else {
            queue.key("queue")
        };
        self.commands.list_size(&key)
    }

    fn reap(
        &self,
        queue: &Queue,
        older_than: i64,
        limit: Option<i64>,
        max_attempts: Option<i64>,
        newer_than: Option<i64>,
    ) -> Result<i64, QueueError> {
        let processing_list = queue.key("processing");
        let now = unix_now();
        let cutoff = now - older_than;
        let mut requeued = 0i64;
        let size = self.commands.list_size(&processing_list)?;
        let claims = self.commands.list_range(&processing_list, size, 0)?;

        for pid_val in claims {
            if limit.is_some_and(|l| requeued >= l) {
                break;
            }
            let Some(pid) = pid_val.as_str() else {
                continue;
            };
            let Some(job) = self.get_job(queue, pid)? else {
                self.commands.list_remove(&processing_list, pid)?;
                continue;
            };
            if job.get_timestamp() > cutoff {
                continue;
            }
            if max_attempts.is_some_and(|m| job.get_attempts() >= m)
                || newer_than.is_some_and(|n| job.get_timestamp() < now - n)
            {
                self.commands.list_remove(&processing_list, pid)?;
                self.commands.left_push(&queue.key("dead"), pid)?;
                continue;
            }
            self.requeue(queue, &job)?;
            self.commands.list_remove(&processing_list, pid)?;
            requeued += 1;
        }
        Ok(requeued)
    }
}

impl Consumer for Redis {
    fn receive(&self, queue: &Queue, timeout: i64) -> Result<Option<Message>, QueueError> {
        if self.is_closed() {
            return Ok(None);
        }
        let next = match self.receive.right_pop_array(&queue.key("queue"), timeout) {
            Ok(v) => {
                let attempt = self.state.reconnect_attempt.load(Ordering::SeqCst);
                if attempt > 0 {
                    self.trigger_reconnect_success(queue, attempt);
                }
                self.state
                    .reconnect_backoff_ms
                    .store(RECONNECT_BACKOFF_MS, Ordering::SeqCst);
                self.state.reconnect_attempt.store(0, Ordering::SeqCst);
                v
            }
            Err(e) if e.is_redis() => {
                if self.is_closed() {
                    return Ok(None);
                }
                let attempt = self.state.reconnect_attempt.fetch_add(1, Ordering::SeqCst) + 1;
                self.receive.close();
                let cap = self.state.reconnect_backoff_ms.load(Ordering::SeqCst);
                let sleep_ms = jitter(cap);
                self.trigger_reconnect(queue, &e, attempt, sleep_ms);
                thread::sleep(Duration::from_millis(sleep_ms));
                let next_backoff = RECONNECT_MAX_BACKOFF_MS.min(cap.saturating_mul(2));
                self.state
                    .reconnect_backoff_ms
                    .store(next_backoff, Ordering::SeqCst);
                return Ok(None);
            }
            Err(e) => return Err(e),
        };

        let Some(mut next_message) = next else {
            return Ok(None);
        };
        if let Some(ts) = next_message.get("timestamp").cloned() {
            let ts_i = match ts {
                Value::Number(n) => n.as_i64().unwrap_or(0),
                Value::String(s) => s.parse().unwrap_or(0),
                _ => 0,
            };
            next_message["timestamp"] = json!(ts_i);
        }
        let message = Message::from_value(&next_message);
        let pid = message.get_pid().to_owned();
        let job_key = format!("{}.jobs.{}.{}", queue.namespace, queue.name, pid);
        self.receive
            .set_array(&job_key, &next_message, queue.job_ttl)?;
        self.receive.left_push(&queue.key("processing"), &pid)?;
        self.receive
            .increment(&format!("{}.stats.{}.total", queue.namespace, queue.name))?;
        self.receive.increment(&format!(
            "{}.stats.{}.processing",
            queue.namespace, queue.name
        ))?;
        Ok(Some(message))
    }

    fn commit(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        let pid = message.get_pid();
        self.commands
            .remove(&format!("{}.jobs.{}.{}", queue.namespace, queue.name, pid))?;
        self.commands
            .increment(&format!("{}.stats.{}.success", queue.namespace, queue.name))?;
        self.commands.list_remove(&queue.key("processing"), pid)?;
        self.commands.decrement(&format!(
            "{}.stats.{}.processing",
            queue.namespace, queue.name
        ))?;
        Ok(())
    }

    fn reject(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        let pid = message.get_pid();
        self.commands.left_push(&queue.key("failed"), pid)?;
        self.commands
            .increment(&format!("{}.stats.{}.failed", queue.namespace, queue.name))?;
        self.commands.list_remove(&queue.key("processing"), pid)?;
        self.commands.decrement(&format!(
            "{}.stats.{}.processing",
            queue.namespace, queue.name
        ))?;
        Ok(())
    }

    fn close(&self) {
        self.state.closed.store(true, Ordering::SeqCst);
    }

    fn as_publisher(&self) -> Option<&dyn Publisher> {
        Some(self)
    }
}

pub(crate) fn unix_now() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

pub(crate) fn unix_now_f64() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

/// PHP `uniqid(more_entropy: true)`-shaped id.
pub(crate) fn uniqid() -> String {
    static COUNTER: AtomicU64 = AtomicU64::new(1);
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    let n = COUNTER.fetch_add(1, Ordering::Relaxed);
    format!(
        "{:08x}{:05x}.{:08x}",
        now.as_secs(),
        now.subsec_micros() % 0x1_00000,
        n as u32
    )
}

fn jitter(max_inclusive: u64) -> u64 {
    if max_inclusive == 0 {
        return 0;
    }
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| u64::from(d.subsec_nanos()))
        .unwrap_or(1);
    nanos % (max_inclusive + 1)
}
